### ISSUE TITLE
Auto complete option in Spyder

### ISSUE THREAD

**Issue by abhijithvnair94** [spyder-ide/spyder#8262]
I am using spyder3. Previously it is possible to do autocomplete using TAB key. But now its not working properly. Kindly help me.

**Comment by jnsebgosselin** [comment]
Please, describe a little bit what it does instead of what you are expecting and fill out the info below.

## Versions


* Spyder version: 
* Python version: 
* Qt version: 
* PyQt version: 
* Operating System name/version:

**Comment by outatbobs** [comment]
Thanks, I like what your doing but my mail box is over flowing with your emails. I can't see my personal mail so I need to stop your sending of your updates. I'll keep an eye on the development of your product which I think is great. Thank you, Bob Sakall

**Comment by abhijithvnair94** [comment]
Earlier when I use TAB key to autocomplete the commands in Spyder Editor Console. But now it is not working properly
 Spyder version: 3.3.1
Python Version: 3.6.6
PyQt Version:4.4.1
Os : Ubuntu 18.04 LTS

**Comment by jnsebgosselin** [comment]
Ok, thank you for the infos. And if you check the option `Tab always indent`, does it work with `Ctrl + Space`?

I'm trying to figure out if this is problem with the `Tab` shortcut or if it's with the auto-completion feature.

![image](https://user-images.githubusercontent.com/10170372/48673241-6bc56180-eb0d-11e8-908f-e9dc46ed4fca.png)

**Comment by abhijithvnair94** [comment]
No sir, I tried with the above mentioned option. After that I tried to use `Ctrl + Space` . But that also not working. I am thinking that this is not the problem of `Tab`.

**Comment by jnsebgosselin** [comment]
Ok, thank you very much @abhijithvnair94 

@ccordoba12 @CAM-Gerlach I think this one is for you guys.

**Comment by abhijithvnair94** [comment]
What  I have to do sir? Kindly explain

**Comment by jnsebgosselin** [comment]
I'll let @CAM-Gerlach answer this one. He is more knowledgeable than me on this topic.

**Comment by CAM-Gerlach** [comment]
> PyQt Version:4.4.1

We don't support PyQt<5.6 as of Spyder >3.2.8, as it has been end of life for many years. Please update to 5.6.x or 5.9.x. This probably isn't directly causing your problem but you are likely to run into other serious issues.

As for your main problem, sorry its not working. This is essentially a duplicate of e.g. #5782 and its many friends, and unfortunately is a very common issue. We're in the process of implementing a a long-term solution in the soon to be released Spyder 4, as we've replaced our completion infrastructure with the Language Server Protocol architecture used by VSCode, Atom, etc. and are currently working hard to make it much more stable and robust. Here are some things you can try in the meantime (in order):

* Restart Spyder. Sometimes, completion will stop working after leaving it open for a while, locking the screen, sleeping, etc.
* Reset your preferences (``spyder --reset``)
* Uninstall ``enum34`` in case it is installed.
* Reboot your machine
* Update Jedi, rope and parso
* Downgrade your packages (particularly those) to the versions in the latest Anaconda release (5.3.0)
* Install Spyder in a clean conda/virtual/v env
* (Re)-install Anaconda

**Comment by vidassadauskas** [comment]
Restarting Spyder (to be more precise - quiting and starting again) helped. Thank you, @CAM-Gerlach, for advice.

**Comment by palashmondal351** [comment]
To enable your autocomplete type the following command. 
1. sudo pip install rope
2. sudo pip install rope_py3k
Then restart your program it will work.

**Comment by CAM-Gerlach** [comment]
The original issue was fixed and closed nearly a year ago, and even the individual who just commented (likely for an unrelated problem) also said their problem was resolved, so I have no idea who you're responding to or why you are posting the above, never mind why you possess such terribly misplaced confidence in the above advice being so certain to fix any and all of a large array of potential causes for why autocomplete may not work.

*TL;DR:* Do **NOT** do any of the above if you don't want to break your install while not fixing the problem.

As a former Spyder core developer, I can categorically state to anyone else reading this that **you should not, under any circumstances, follow the above advice**, as it is far more likely to break your install and/or lead to even more problems down the road than fix whatever your particular problem is, as it has numerous problems.

First off, given the number of issues that can cause autocomplete to stop working, the above even in the best case is only likely to fix a tiny, tiny minority of them, if any at all.

Second, given its use of `sudo`, it may be modifying the system Python install rather than that in a virtualenv, Anaconda environment, etc. On Windows it will simply not work at all and result in an error, while on Macs it either won't work or cause an unexpected result depending on `which pip`. On Linux, it either will install the packages using the system Python (which is almost certainly not what you wan) or install them as `root` with whatever Python `pip` points to, which at least for a global Python install will be in a Python 2 rather than Python 3 environment, as the vast majority of users are using nowadays. Regardless, all of this is a huge mess and a very bad idea.

Third, you are installing two mutually incompatible packages, `rope` and `rope_py3k`; `rope_py3k` is a fork of `rope` that is Py3-only that has not been maintained for many years, and will break if the user installs it on a Python 2 environment, while `rope` is the original, maintained and now Python 3 compatible package. Installing both could leak to a conflict, now or down the road, so it is not recommended, and if `rope_py3k` shadows `rope` the user will be using an extremely out of date and unmaintained package.

Finally, given the great majority of Spyder users use `conda`, not `pip`, and already have `rope` installed if they have installed Spyder or are using the `base` environment, the above command will install not one but two duplicate versions of rope with `pip` on top of the existing `conda` installs, which is a surefire way to cause no end of trouble particularly if done into the default `base` env. Furthermore, even if not, an install via `pip` should be avoided in favor of `conda` whenever possible to avoid all manner of tricky installation issues.

Ergo, under no circumstances should anyone execute any of the above commands, as they will almost certainly not fix your problem (as it is a very-specific, temporary fix to a only one of dozens/hundreds of problems that can cause this issue, and furthermore only for a very specific, uncommon and very not-recommended environment setup) and is far more likely to break your install in any one of numerous ways.

While an attempt to provide advice is laudable, please ensure you fully, completely and thoroughly understand exactly what you are suggesting people do (_particularly_ if they involve any one of `sudo`, install commands and/or `pip`, all things strongly not recommended or supported by the Spyder team), what environments and specific root problems it will work on, and what the consequences are, and explain those in your post before declaring categorically that it will fix the very large, general problem you claim it to. Thanks.

**Comment by llsouza** [comment]
> 
> 
> To enable your autocomplete type the following command.
> 
>     1. sudo pip install rope
> 
>     2. sudo pip install rope_py3k
>        Then restart your program it will work.

DO NOT DO THIS! SEE REPLY BY CAM-Gerlach.  This bad advise has been displayed prominently on searches.

**Comment by shahnawazm786** [comment]
![image](https://user-images.githubusercontent.com/49604292/90748874-5be53d80-e2f0-11ea-8489-b238a8bdadc5.png)
i am not able get the intelli sense  in spyder.
even though the both is ticked.
Thanks

**Comment by shahnawazm786** [comment]
> Ctrl + Space
same with me. nothing is happening with spyder.

**Comment by Ranj1tha** [comment]
Even am facing same issue, can anyone got any solution?

**Comment by solocazzimiei** [comment]
Hi . I've installed Spyder through pip commad : its will work fine if would do also the code completion....neghter `TAB` nor `Ctrl+Space` activate the function.. system characteristics: Spyder 5.1.5|Python 3.9.7 64-bit | Qt 5.12.10 | PyQt5 5.12.3 | Windows 10

**Comment by CAM-Gerlach** [comment]
Hey, sorry about that. Did you download and install Spyder via our recommended standalone installers (as opposed to Anaconda, WinPython, pip etc)? Since this can vary widely from case to case, could you post a GIF or short video (or if that's not possible, at least a screenshot) showing the specifics of your attempted completion? Does it not work in the console, editor or both? Did you try completion on basic standard library modules (e.g. `os`) as well as third party ones? Thanks!

**Comment by solocazzimiei** [comment]
thanks for quick interest ! I remember that on version 3.9 installed from Anaconda, the auto-completion was working. Now that Spyder were installed through pip on line download I can't see any sign of editor's help with any library.  In tools-->Preferences-->General, the flags are set to "on"

**Comment by CAM-Gerlach** [comment]
> Now that Spyder were installed through pip on line download I can't see any sign of editor's help with any library.

I'm not 100% sure which specific install method you're describing here, but it seems likely it is due to an install issue. In our [installation guide](https://docs.spyder-ide.org/current/installation.html), could you point out which section matches the approach you followed to install Spyder? We highly recommend our standalone installers, and if not that Anaconda, to avoid these sorts of problems, which are usually caused by installation or third-party dependency issues that using our dedicated installer helps avoid. Thanks!

**Comment by belongstodzire** [comment]
I used `sudo apt-get` to install spyder as a stand-alone installer. I too am unable to use autocomplete feature. 
Spyder 3.3.6
Python 3.8.10
![image](https://user-images.githubusercontent.com/63529288/153701973-85a17955-bf7f-4ace-9a8d-e6af702c7c76.png)
![image](https://user-images.githubusercontent.com/63529288/153702005-06a17f3c-3e0f-4db3-a674-1728b089d9fa.png)

**Comment by CAM-Gerlach** [comment]
@belongstodzire Spyder 3.3.6 is nearly 3 years out of date and uses a completely different, less functional and reliable autocompletion system from the current version, Spyder 5.2.2. Furthermore, per our [installation guide](https://docs.spyder-ide.org/current/installation.html), using third-party Spyder builds from your distro's package manager is not recommended or officially supported, are not maintained by the Spyder team, may be years out of date and could contain bugs and issues separate from those in officially supported Spyder builds.

On Linux, if you're a beginning or intermediate user, we suggest using Spyder as packaged with Anaconda/Miniconda/Miniforge, or if you're an advanced user, installing it with `pip` in a fresh virtualenv/venv. See the above-linked installation guide for details on those and other install methods. You can also ask your distro's package maintainers to update their obselete, EoL version of Spyder to something supported; we've reached out to other distro maintainers in the past but despite making initial contact and many followup attempts, have been unable to reliably get ahold of those in charge of the Debian/Ubuntu packages, even to update the package description and metadata which is nearly a decade and a half out of date.

Best of luck!

**Comment by belongstodzire** [comment]
> @belongstodzire Spyder 3.3.6 is nearly 3 years out of date and uses a completely different, less functional and reliable autocompletion system from the current version, Spyder 5.2.2. Furthermore, per our [installation guide](https://docs.spyder-ide.org/current/installation.html), using third-party Spyder builds from your distro's package manager is not recommended or officially supported, are not maintained by the Spyder team, may be years out of date and could contain bugs and issues separate from those in officially supported Spyder builds.
> 
> On Linux, if you're a beginning or intermediate user, we suggest using Spyder as packaged with Anaconda/Miniconda/Miniforge, or if you're an advanced user, installing it with `pip` in a fresh virtualenv/venv. See the above-linked installation guide for details on those and other install methods. You can also ask your distro's package maintainers to update their obselete, EoL version of Spyder to something supported; we've reached out to other distro maintainers in the past but despite making initial contact and many followup attempts, have been unable to reliably get ahold of those in charge of the Debian/Ubuntu packages, even to update the package description and metadata which is nearly a decade and a half out of date.
> 
> Best of luck!

Thanks a lot @CAM-Gerlach for the reply. I understood I should have installed it through Anaconda. However, laziness prompted me to do a ```sudo apt-get``` . I have understood the problem and would rectify it asap.
